### ISSUE TITLE
HLS Video zoom-in when select a video track with higher resolution, and zoom-out when select a video track with lower resolution

### ISSUE THREAD

**Issue by amarnehSoft** [androidx/media#1831]
### Version

Media3 main branch

### More version details

commit: d2fb779929f05b01b52b78873d39367d72440ee0

### Devices that reproduce the issue

Pixel8 running android 15 (emulator)

### Reproducible in the demo app?

Yes

### Reproduction steps

1. in short-forms demo project, add this sample HLS video url "https://demo.unified-streaming.com/k8s/features/stable/video/tears-of-steel/tears-of-steel.ism/.m3u8" to the list in MediaItemDatabase class.
2. wait a little bit, until another track get selected based on the network strength, (will select lower resolution if network is low, and will select higher resolution if network is good enough)

### Expected result

the player switch tracks correctly and change the resolution without ui issues

### Actual result

the player shows the new track zoomed in if the track is a higher resolution, or will show it zoomed out if the track is lower resolution
![Screenshot_20241024_230734](https://github.com/user-attachments/assets/60610034-7a98-4556-8b33-5095c121b785)
![Screenshot_20241024_232030](https://github.com/user-attachments/assets/76295397-fd6e-4378-8521-2450fa8040f3)

### Media

https://demo.unified-streaming.com/k8s/features/stable/video/tears-of-steel/tears-of-steel.ism/.m3u8

### Bug Report

- [ ] You will email the zip file produced by `adb bugreport` to android-media-github@google.com after filing this issue.

**Comment by marcbaechinger** [comment]
The shortform demo isn't a general purpose video player app but tailored to test preloading. For this reason the UI is specific to videos in portrait mode that fill the entire screen. This allows us to visually test preloading. 

One thing I usually do is scrolling just very few pixels up and down to see only a very small fraction of the next surface in the view pager to verify it's preloaded properly. That's why the video surface should fill the entire activity.

So I think this behaviour is intended and we specifically for this purpose added the attribute

```
    app:resize_mode="fill"
```

to the `PlayerView` element to avoid a letterbox display in the activity.

If you remove that attribute from [`layout/media_item_view_pager.xml`](https://github.com/androidx/media/blob/release/demos/shortform/src/main/res/layout/media_item_view_pager.xml#L29) then your stream is displayed without distortion. You can also see that for the other samples there is a black frame above and below the video (letterboxing) which is what that settings tries to avoid.

**Comment by amarnehSoft** [comment]
sorry i might not describe the issue clearly, the issue isn't in wrong aspect ration, the issue is that when i select another track (lower resolution for example) the new track will shown in a smaller frame at top left, while the previous track freezes and occupies the old size, please check the attached screenshot, it shows two tracks overlaps each other, the old track shown on entire screen and it's freeze, and the new track is on top of the old track at top left with smaller size.
And the issue isn't only in short-forms project, i can reproduce this issue on any player runs HLS, for example demo-compose project.
![Screenshot_20241024_232030](https://github.com/user-attachments/assets/c73d07d9-f087-4758-b5a9-734fc9812572)

**Comment by marcbaechinger** [comment]
I tested on Pixel 6, 7 and 8 and I can not reproduce on a real device. I can repro on a Pixel 7 and 8 emulator consistently though.

When you say you can repro on `Pixel8 running android 15 (emulator)` does this mean you haven't seen that on a real devices? Or do you mean that it is specific to API 35 on  an emulator?

**Comment by amarnehSoft** [comment]
after testing on real device, seems the issue is only on emulators, and it's working fine on real devices, so i will close the issue, thanks.

**Comment by Senne021** [comment]
@marcbaechinger @amarnehSoft I am currently facing the same issue and have delved deeper into its causes. The problem occurs on all devices we've tested so far, including both emulators and real devices. The root cause seems to be a combination of the HLS video, the Media Codec, and hardware acceleration. Here’s what I’ve discovered so far:

Depending on the initial resolution, the displayed image becomes smaller or larger when the video quality changes.
The issue only occurs with videos in the AVC format.
If I manually select the Media Codec and choose one without hardware acceleration, the problem no longer occurs.
This issue persists for me across all versions of the ExoPlayer.

**Comment by marcbaechinger** [comment]
Thanks for your report!

How can we repro this? Can you repro on the demo app? Are you using compose or `PlayerView`? Please provide some repro steps and some more details to make this actionable for us.

Can you give us a bit more detail how you build the UI? Best would be if you can repro in the demo app. If this is not possible, then we'd need to know how the UI is built and what is, for instance, sizing the surface. It would for instance be interesting to know what the dimension of the surface is in the screenshots above.

This this work before? What did you change to make that made it render wrong? Anything like that is helpful for us. Thanks for the test stream

> This issue persists across all versions of the ExoPlayer.

Can you be specific around versions? If you say all version I would be surprised we haven't heard about this and it makes me think it may be a problem specific to that UI? I'm guessing based on little information availabel. Don't get me wrong.

If it's about codec things like you suggest, then the API level would be interesting to know. If you say `all devices` I wonder why we haven't heard the community yell. Can you be more detailed about a common combination of device/API/ExoPlayer version that we can easily get our hand on to repro?

**Comment by Senne021** [comment]
Hi @marcbaechinger,
I’ve created a small sample app, which you can find [here](https://github.com/Senne021/exopplayer_hls). This app demos the issue with HLS video playback, where a quality change results in the surface size being incorrectly adjusted.

I’ve tested this sample app on various devices, and it appears to be a problem related to the media codec. The issue occurs on all my simulated devices, as well as my Android TV devices and FireOS devices. Interestingly, it does not happen on my smartphone (S23), which uses a different media codec. This suggests that the problem might stem from a peculiar combination of the media codec, the HLS video format, and the specific environment of smart TVs or simulators.

Would you like more information about the devices and media codecs I used to test this demo app?

**Comment by marcbaechinger** [comment]
> The issue occurs on all my simulated devices, as well as my Android TV devices and FireOS devices.

Interesting. Then this seems to be a problem of the HW codecs on these devices. Specifically when I see how you workaround this with the SW codec in the sample.  I'm unsure whether I can really help with that, but I may find someone from the TV team who can give me a nudge. Thanks for reporting!

Just curious: Is this happening with most HLS streams or is the one above special as it is one out of many that behaves like this? If you take another like the apple sample stream or the like then you can't repro?

**Comment by Senne021** [comment]
> Just curious: Is this happening with most HLS streams or is the one above special as it is one out of many that behaves like this? If you take another like the apple sample stream or the like then you can't repro?

This seems to depend on the video encoding. With AVC streams, this has always happened as soon as the video quality has changed. The problem does not seem to exist with [this](https://hlsjs.video-dev.org/demo/?src=https%3A%2F%2Fdevstreaming-cdn.apple.com%2Fvideos%2Fstreaming%2Fexamples%2Fbipbop_adv_example_hevc%2Fmaster.m3u8&demoConfig=eyJlbmFibGVTdHJlYW1pbmciOnRydWUsImF1dG9SZWNvdmVyRXJyb3IiOnRydWUsInN0b3BPblN0YWxsIjpmYWxzZSwiZHVtcGZNUDQiOmZhbHNlLCJsZXZlbENhcHBpbmciOi0xLCJsaW1pdE1ldHJpY3MiOi0xfQ==) HEVC stream.

**Comment by nisrulz** [comment]
I seem to have hit this issue with Exoplayer at work. 

I tried to summarize the comments and conversation uptil here in this PR with Github Copilot, so it is easier to consume:

---

- Bug: When switching HLS video tracks to a higher or lower resolution, the video would overlap with a frozen previous frame.
- One commenter reported that the issue occurs on emulator and some real devices, especially Android TV and FireOS devices. They found:
  - The bug happens with HLS streams using the AVC codec and hardware acceleration.
  - **Switching to a software codec avoids the problem.**
  - It can be reproduced across all ExoPlayer versions tested.
  - The bug depends on the codec and device (e.g., not seen on a Samsung S23 smartphone).
  - The problem does not occur with HEVC streams, only AVC.
- The maintainers acknowledged it seems to be a hardware codec-specific issue, primarily affecting TV and emulator environments, and may require help from the TV team.

In summary: The issue is codec- and device-specific, mainly affecting hardware-accelerated AVC HLS playback on certain emulators and smart TV devices, but not universally reproducible on all phones or with all codecs (HEVC seems unaffected). Workarounds include using software decoding. The maintainers are aware and may escalate to the TV team for deeper investigation.

---

Follow up question:
- What would be a possible workaround approach to handle this issue?

> I tried to recreate the Exoplayer and it seems to fix the issue, but I would like to not have to recreate Exoplayer altogether.

Any suggestion/pointers would be helpful. Thank you!

**Comment by ELOUFIR** [comment]
I can confirm what @nisrulz  reported.
I reproduced the same issue in my **Amazon Tv Firestick TV** and also using the Android Studio emulator (Windows 11).

**Steps to reproduce:**

1. Play this live news channel: https://live.alarabiya.net/alarabiapublish/alarabiya.smil/playlist.m3u8
2. Wait for the track to change automatically, or change it manually.

I will try to make a sample code when I find the time.

Emulator Properties:
````
avd.ini.displayname              Pixel Tablet (Edited 7 inch) API 36
avd.ini.encoding                 UTF-8
AvdId                            Pixel_Tablet_Edited_7_inch_API_36
disk.dataPartition.size          6442450944
fastboot.chosenSnapshotFile      
fastboot.forceChosenSnapshotBoot no
fastboot.forceColdBoot           no
fastboot.forceFastBoot           yes
hw.accelerometer                 yes
hw.arc                           false
hw.audioInput                    yes
hw.battery                       yes
hw.camera.back                   virtualscene
hw.camera.front                  none
hw.cpu.ncore                     4
hw.device.hash2                  MD5:155fe64c0164b503ef4797be44f519c2
hw.device.manufacturer           User
hw.device.name                   Pixel Tablet (Edited 7 inch)
hw.dPad                          no
hw.gps                           no
hw.gpu.enabled                   yes
hw.gpu.mode                      auto
hw.gyroscope                     yes
hw.initialOrientation            landscape
hw.keyboard                      yes
hw.lcd.density                   320
hw.lcd.height                    1200
hw.lcd.width                     1920
hw.mainKeys                      no
hw.ramSize                       2048
hw.sdCard                        yes
hw.sensors.light                 yes
hw.sensors.magnetic_field        yes
hw.sensors.orientation           yes
hw.sensors.pressure              yes
hw.sensors.proximity             no
hw.trackBall                     no
image.sysdir.1                   system-images\android-36\google_apis\x86_64\
PlayStore.enabled                false
runtime.network.latency          none
runtime.network.speed            full
showDeviceFrame                  yes
skin.dynamic                     yes
tag.display                      Google APIs
tag.displaynames                 Google APIs
tag.id                           google_apis
tag.ids                          google_apis
target                           android-36
vm.heapSize                      128
````
